_posts/en/newsletters/2019-05-14-newsletter.md

```diff
@@ -0,0 +1,485 @@
+
+Under Taproot, the hashing step will be skipped and so the bech32
+address will contain the pubkey directly, with one small change.
+Currently, 33-byte Bitcoin-style pubkeys are encoded to start with either
+a 0x02 or 0x03 to allow validators to reconstruct the key's Y-coordinate
+on the secp256k1 elliptic curve; in bip-taproot, the value of this byte
+is reduced by two so that 0x02 becomes 0x00 and 0x03 becomes 0x01,
+although the meaning is exactly the same.  Also the witness version is
+changed from the `0` used for P2WPKH/P2WSH to a `1`.
+
+| Object           | Operation                                                       | Example result |
+| Private key      | read 32 bytes from [CSPRNG][], or using [BIP32][] HD derivation | 0xFIXME |
+| Public key       | point(0xFIXME), or using [BIP32][] HD public derivation         | 0x02FIXME |
+| Alter key prefix | (key[0] - 2),key[1:33])                                         | 0x00FIXME |
+| Address          | bech32.encode(1, 0x00FIXME)                                     | bc1pFIXME |
+
+Here's an example of existing addresses compared to an example taproot
+address.
+
+    p2pkh FIXME
+    p2sh FIXME
+    p2wpkh FIXME
+    p2wsh FIXME
+    taproot FIXME
+
+Spending from P2PKH or P2WPKH requires the spender include their public
+key in their input.  For Taproot, the public key was provided in the
+UTXO being spent, so several vbytes are saved by not including it again.
+The spend must also include a signature; this is a Schnorr-style
+signature as defined by [bip-schnorr][] with an optional sighash byte
+appended.  If the default sighash is used, the signature is 64 bytes (16
+vbytes); if a non-default is used, it is 65 bytes (16.25
+vbytes[^vbytes-decimal]).  Overall, this makes the cost to create and
+spend a Taproot single-sig output about 1% more expensive than P2WPKH.
+However, the costs are distributed differently---it costs more to create
+a Taproot output than a P2WPKH output and less to spend it, which may
+help contribute towards keeping the size of the UTXO set manageable.
+
+                               FIXME:tablify
+                               Vbytes
+                           P2PKH          P2WPKH               Taproot
+    scriptPubKey           24             22                   33
+    scriptSig              1+33+1+72=107  0                    0
+    witness                0              (1+33+1+72)/4=26.75  (1+64)/4=16.25
+    TOTAL                  131            48.75                49.25
+
+Besides the change from ECDSA to Schnorr for the signature algorithm,
+there are a few important (but easy to implement) changes to the
+transaction digest---the hash that a signature commits to in order to
+prove the transaction is an authorized spend of a UTXO.
+
+Most notably, the hash used goes from the legacy protocol's
+double-SHA256 (SHA256d) to a single SHA256 operation.  The extra hashing
```

Comment:
I haven't measured, but some napkin math says that if all 412 million txes on the block chain have an average of 2 inputs each with 1 signature, this could save a 10 to 20 minutes off of IBD.  However, the transaction digest formula now includes hashes in other places, so I think the total number of hashes increases (although the amount of data being hashed decreases), which makes it beyond the capacity of my napkin to clearly evaluate.